probe-rs/src/core/registers.rs

```diff
@@ -293,151 +351,220 @@ impl<T> RegisterValueResultExt<T> for Result<T, Infallible> {
-    pub fn other(&self) -> impl Iterator<Item = &RegisterDescription> {
-        self.other.iter()
+    pub fn other(&self) -> impl Iterator<Item = &CoreRegister> {
+        self.0
+            .iter()
+            .filter(|r| matches!(r.role, Some(RegisterRole::Other(_))))
     }
 
     /// Find an architecture specific register by name
-    pub fn other_by_name(&self, name: &str) -> Option<&RegisterDescription> {
-        self.other.iter().find(|r| r.name == name)
+    pub fn other_by_name(&self, name: &str) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .find(|r| matches!(r.role, Some(RegisterRole::Other(other_name)) if other_name == name))
+            .ok_or_else(|| {
+                Error::GenericCoreError(format!(
+                    "Argument register {name:?} not found. Please report this as a bug."
+                ))
+            })
     }
 
     /// The fpu status register.
-    pub fn fpscr(&self) -> Option<&RegisterDescription> {
-        self.fp_status
+    pub fn fpsr(&self) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .find(|r| r.role == Some(RegisterRole::FloatingPointStatus))
+            .ok_or_else(|| {
+                Error::GenericCoreError(
+                    "No FPU status register found. Please report this as a bug.".to_string(),
+                )
+            })
     }
 
     /// Returns an iterator over the descriptions of all the registers of this core.
-    pub fn fpu_registers(&self) -> Option<impl Iterator<Item = &RegisterDescription>> {
-        self.fp_registers.map(|r| r.iter())
+    pub fn fpu_registers(&self) -> Option<impl Iterator<Item = &CoreRegister>> {
+        let mut fpu_registers = self
+            .0
+            .iter()
+            .filter(|r| r.role == Some(RegisterRole::FloatingPoint))
+            .peekable();
+        if fpu_registers.peek().is_some() {
+            Some(fpu_registers)
+        } else {
+            None
+        }
     }
 
     /// Returns the nth fpu register.
     ///
     /// # Panics
     ///
     /// Panics if the register at given index does not exist.
-    pub fn fpu_register(&self, index: usize) -> Option<&RegisterDescription> {
-        self.fp_registers.map(|r| &r[index])
+    pub fn fpu_register(&self, index: usize) -> &CoreRegister {
+        self.get_fpu_register(index).unwrap()
     }
 
     /// Returns the nth fpu register if it is exists, `None` otherwise.
-    pub fn get_fpu_register(&self, index: usize) -> Option<&RegisterDescription> {
-        self.fp_registers.and_then(|r| r.get(index))
+    pub fn get_fpu_register(&self, index: usize) -> Result<&CoreRegister, Error> {
+        self.0
+            .iter()
+            .filter(|r| r.role == Some(RegisterRole::FloatingPoint))
+            .nth(index)
+            .ok_or_else(|| {
+                Error::GenericCoreError(format!(
+                    "Argument register {index:?} not found. Please report this as a bug."
```

Comment:
This is not necessarily bug, if the register doesn't exist on the chip.

Reply:
Agreed. Consider it Done :)